File: frontend/src/engine/news2Engine.ts
import { Vitals, NEWS2Score, RiskTier } from '../types/hdims';

/**
 * Official NHS / Royal College of Physicians NEWS2 (National Early Warning Score) engine.
 * Computes clinical risk and deterioration state from real-time patient physiological signals.
 */
export function calculateNEWS2(vitals: Vitals): NEWS2Score {
  let rrScore = 0;
  if (vitals.respirationRate <= 8) rrScore = 3;
  else if (vitals.respirationRate >= 9 && vitals.respirationRate <= 11) rrScore = 1;
  else if (vitals.respirationRate >= 12 && vitals.respirationRate <= 20) rrScore = 0;
  else if (vitals.respirationRate >= 21 && vitals.respirationRate <= 24) rrScore = 2;
  else if (vitals.respirationRate >= 25) rrScore = 3;

  let spo2Score = 0;
  if (vitals.spo2 <= 91) spo2Score = 3;
  else if (vitals.spo2 >= 92 && vitals.spo2 <= 93) spo2Score = 2;
  else if (vitals.spo2 >= 94 && vitals.spo2 <= 95) spo2Score = 1;
  else spo2Score = 0;

  const o2Score = vitals.oxygenSupplement ? 2 : 0;

  let bpScore = 0;
  if (vitals.systolicBp <= 90) bpScore = 3;
  else if (vitals.systolicBp >= 91 && vitals.systolicBp <= 100) bpScore = 2;
  else if (vitals.systolicBp >= 101 && vitals.systolicBp <= 110) bpScore = 1;
  else if (vitals.systolicBp >= 111 && vitals.systolicBp <= 219) bpScore = 0;
  else bpScore = 3;

  let hrScore = 0;
  if (vitals.pulseRate <= 40) hrScore = 3;
  else if (vitals.pulseRate >= 41 && vitals.pulseRate <= 50) hrScore = 1;
  else if (vitals.pulseRate >= 51 && vitals.pulseRate <= 90) hrScore = 0;
  else if (vitals.pulseRate >= 91 && vitals.pulseRate <= 110) hrScore = 1;
  else if (vitals.pulseRate >= 111 && vitals.pulseRate <= 130) hrScore = 2;
  else hrScore = 3;

  let consciousnessScore = 0;
  if (vitals.consciousness !== 'ALERT') consciousnessScore = 3;

  let tempScore = 0;
  if (vitals.temperature <= 35.0) tempScore = 3;
  else if (vitals.temperature >= 35.1 && vitals.temperature <= 36.0) tempScore = 1;
  else if (vitals.temperature >= 36.1 && vitals.temperature <= 38.0) tempScore = 0;
  else if (vitals.temperature >= 38.1 && vitals.temperature <= 39.0) tempScore = 1;
  else tempScore = 2;

  const subscores = {
    respirationRate: rrScore,
    spo2: spo2Score,
    oxygenSupplement: o2Score,
    systolicBp: bpScore,
    pulseRate: hrScore,
    consciousness: consciousnessScore,
    temperature: tempScore,
  };

  const total = rrScore + spo2Score + o2Score + bpScore + hrScore + consciousnessScore + tempScore;

  const hasExtremeSingleScore = Object.values(subscores).some((val) => val === 3);

  let tier: RiskTier = 'LOW';
  let tierLabel = 'Tier 1 — Stable';
  let triggersEscalation = false;
  let clinicalAction = 'Routine 4-6 hourly nursing monitoring. Standard ward care.';

  if (total >= 7) {
    tier = 'CRITICAL';
    tierLabel = 'Tier 3 — Critical Deterioration';
    triggersEscalation = true;
    clinicalAction = 'IMMEDIATE EMERGENCY: Pre-provision Rapid Response & On-Call Specialist. Code Blue Standby.';
  } else if (total >= 5 || hasExtremeSingleScore) {
    tier = 'MEDIUM';
    tierLabel = 'Tier 2 — Acute Physiological Decline';
    triggersEscalation = true;
    clinicalAction = 'Urgent Clinical Review: Alert Attending Physician. Hourly monitoring. Oxygen titration.';
  }

  return {
    total,
    tier,
    tierLabel,
    subscores,
    triggersEscalation,
    clinicalAction,
  };
}

File: backend/app/news2_engine.py
from typing import Dict, Any, Tuple

def calculate_news2(
    respiration_rate: int,
    spo2: int,
    oxygen_supplement: bool,
    systolic_bp: int,
    pulse_rate: int,
    consciousness: str,
    temperature: float
) -> Tuple[int, str, str, Dict[str, int]]:
    """
    Computes deterministic National Early Warning Score 2 (NEWS2)
    based on Royal College of Physicians UK clinical standard.
    Returns: (total_score, tier, clinical_action, subscores)
    """
    subscores = {}

    # 1. Respiration Rate
    if respiration_rate <= 8:
        subscores["respiration"] = 3
    elif 9 <= respiration_rate <= 11:
        subscores["respiration"] = 1
    elif 12 <= respiration_rate <= 20:
        subscores["respiration"] = 0
    elif 21 <= respiration_rate <= 24:
        subscores["respiration"] = 2
    else:
        subscores["respiration"] = 3

    # 2. SpO2 Oxygen Saturation (Scale 1 standard)
    if spo2 <= 91:
        subscores["spo2"] = 3
    elif 92 <= spo2 <= 93:
        subscores["spo2"] = 2
    elif 94 <= spo2 <= 95:
        subscores["spo2"] = 1
    else:
        subscores["spo2"] = 0

    # 3. Supplemental Oxygen
    subscores["oxygen"] = 2 if oxygen_supplement else 0

    # 4. Systolic Blood Pressure
    if systolic_bp <= 90:
        subscores["systolic_bp"] = 3
    elif 91 <= systolic_bp <= 100:
        subscores["systolic_bp"] = 2
    elif 101 <= systolic_bp <= 110:
        subscores["systolic_bp"] = 1
    elif 111 <= systolic_bp <= 219:
        subscores["systolic_bp"] = 0
    else:
        subscores["systolic_bp"] = 3

    # 5. Pulse / Heart Rate
    if pulse_rate <= 40:
        subscores["pulse"] = 3
    elif 41 <= pulse_rate <= 50:
        subscores["pulse"] = 1
    elif 51 <= pulse_rate <= 90:
        subscores["pulse"] = 0
    elif 91 <= pulse_rate <= 110:
        subscores["pulse"] = 1
    elif 111 <= pulse_rate <= 130:
        subscores["pulse"] = 2
    else:
        subscores["pulse"] = 3

    # 6. Consciousness (ACVPU)
    c_upper = consciousness.upper()
    if c_upper == "ALERT":
        subscores["consciousness"] = 0
    else:
        subscores["consciousness"] = 3

    # 7. Temperature
    if temperature <= 35.0:
        subscores["temperature"] = 3
    elif 35.1 <= temperature <= 36.0:
        subscores["temperature"] = 1
    elif 36.1 <= temperature <= 38.0:
        subscores["temperature"] = 0
    elif 38.1 <= temperature <= 39.0:
        subscores["temperature"] = 1
    else:
        subscores["temperature"] = 2

    total = sum(subscores.values())

    # Risk Tier classification
    if total >= 7 or any(s == 3 for s in subscores.values() if s == 3 and total >= 5):
        tier = "CRITICAL"
        action = "EMERGENCY: Immediate specialist bedside review. Automatic 45-min JIT token pre-provisioning."
    elif 5 <= total <= 6 or any(s == 3 for s in subscores.values()):
        tier = "MEDIUM"
        action = "URGENT ALERT: Ward doctor review within 30 minutes. Continuous telemetry active."
    else:
        tier = "LOW"
        action = "STABLE: Routine 4-6 hourly monitoring. Zero standing access beyond baseline nurse."

    return total, tier, action, subscores
